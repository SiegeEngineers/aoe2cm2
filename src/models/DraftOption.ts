interface ImageUrls {
    readonly unit: string;
    readonly emblem: string;
    readonly animated_left: string;
    readonly animated_right: string;
}

enum Name {
    HIDDEN_PICK = "HIDDEN_PICK",
    HIDDEN_BAN = "HIDDEN_BAN",
    HIDDEN_SNIPE = "HIDDEN_SNIPE",
    HIDDEN_STEAL = "HIDDEN_STEAL",
    HIDDEN = "HIDDEN",
    RANDOM = "RANDOM",
}

class DraftOption {

    public static readonly HIDDEN_PICK: DraftOption = new DraftOption(Name.HIDDEN_PICK);
    public static readonly HIDDEN_BAN: DraftOption = new DraftOption(Name.HIDDEN_BAN);
    public static readonly HIDDEN_SNIPE: DraftOption = new DraftOption(Name.HIDDEN_SNIPE);
    public static readonly HIDDEN_STEAL: DraftOption = new DraftOption(Name.HIDDEN_STEAL);
    public static readonly HIDDEN: DraftOption = new DraftOption(Name.HIDDEN);
    public static readonly RANDOM: DraftOption = new DraftOption(Name.RANDOM);

    public readonly id: string;
    public readonly name: string;
    public readonly imageUrls: ImageUrls;

    constructor(id: string, name: string = id, imageUrls: ImageUrls = DraftOption.defaultImageUrlsForCivilisation(id)) {
        this.id = id;
        this.name = name;
        this.imageUrls = imageUrls;
    }

    public static defaultImageUrlsForCivilisation(name: string): ImageUrls {
        return {
            unit: `/images/civs/${name.toLowerCase()}.png`,
            emblem: `/images/civemblems/${name.toLowerCase()}.png`,
            animated_left: `/images/units-animated/${name.toLowerCase()}-left.apng`,
            animated_right: `/images/units-animated/${name.toLowerCase()}-right.apng`,
        }
    }

}

export default DraftOption;
