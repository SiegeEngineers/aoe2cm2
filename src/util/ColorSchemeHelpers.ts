import {ColorScheme} from "../constants/ColorScheme";

class ColorSchemeHelpers {

    public static getColorSchemeFromLocalStorage(): ColorScheme {
        try {
            const preference = localStorage.getItem('preferred_color_scheme');
            let scheme: ColorScheme = ColorScheme.AUTO;
            Object.keys(ColorScheme).forEach(function (key) {
                if (ColorScheme[key] === preference) {
                    scheme = ColorScheme[key];
                }
            });
            return scheme;
        } catch (e) {
            return ColorScheme.AUTO;
        }
    }

    public static writeNameToLocalStorage(colorScheme: ColorScheme) {
        try {
            localStorage.setItem('preferred_color_scheme', colorScheme);
        } catch (e) {
            console.error("Couldn't save color scheme to local storage", e);
        }
    }

    public static getOSColorSchemePreference(): ColorScheme {
        const mq = window.matchMedia('(prefers-color-scheme: dark)');
        const preference = mq.matches ? ColorScheme.DARK : ColorScheme.LIGHT;
        console.log('OS Preference mode', preference);
        return preference;
    }

    public static changeColorScheme(colorScheme: ColorScheme) {
        if (colorScheme === ColorScheme.AUTO) {
            colorScheme = ColorSchemeHelpers.getOSColorSchemePreference()
        }
        if (colorScheme === ColorScheme.DARK) {
            this.addThemeClassName('has-theme-dark');
        } else {
            this.removeThemeClassName('has-theme-dark');
        }
    }

    public static addThemeClassName(className: string) {
        document.documentElement.classList.add(className);
        document.body.classList.add(className);
    }

    public static removeThemeClassName(className: string) {
        document.documentElement.classList.remove(className);
        document.body.classList.remove(className);
    }

}

export default ColorSchemeHelpers;
