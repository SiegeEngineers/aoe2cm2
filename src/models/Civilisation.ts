import GameVersion from "../constants/GameVersion";
import DraftOption from "./DraftOption";

enum Name {
    BRITONS = "Britons",
    BYZANTINES = "Byzantines",
    CELTS = "Celts",
    CHINESE = "Chinese",
    FRANKS = "Franks",
    GOTHS = "Goths",
    JAPANESE = "Japanese",
    MONGOLS = "Mongols",
    PERSIANS = "Persians",
    SARACENS = "Saracens",
    TEUTONS = "Teutons",
    TURKS = "Turks",
    VIKINGS = "Vikings",

    AZTECS = "Aztecs",
    HUNS = "Huns",
    KOREANS = "Koreans",
    MAYANS = "Mayans",
    SPANISH = "Spanish",

    ITALIANS = "Italians",
    INCAS = "Incas",
    INDIANS = "Indians",
    MAGYARS = "Magyars",
    SLAVS = "Slavs",

    BERBERS = "Berbers",
    ETHIOPIANS = "Ethiopians",
    MALIANS = "Malians",
    PORTUGUESE = "Portuguese",

    BURMESE = "Burmese",
    KHMER = "Khmer",
    MALAY = "Malay",
    VIETNAMESE = "Vietnamese",

    BULGARIANS = "Bulgarians",
    CUMANS = "Cumans",
    LITHUANIANS = "Lithuanians",
    TATARS = "Tatars",

    BURGUNDIANS = "Burgundians",
    SICILIANS = "Sicilians",

    BOHEMIANS = "Bohemians",
    POLES = "Poles",
}

class Civilisation extends DraftOption {

    public static readonly BRITONS: Civilisation = new Civilisation(Name.BRITONS, GameVersion.AOK);
    public static readonly BYZANTINES: Civilisation = new Civilisation(Name.BYZANTINES, GameVersion.AOK);
    public static readonly CELTS: Civilisation = new Civilisation(Name.CELTS, GameVersion.AOK);
    public static readonly CHINESE: Civilisation = new Civilisation(Name.CHINESE, GameVersion.AOK);
    public static readonly FRANKS: Civilisation = new Civilisation(Name.FRANKS, GameVersion.AOK);
    public static readonly GOTHS: Civilisation = new Civilisation(Name.GOTHS, GameVersion.AOK);
    public static readonly JAPANESE: Civilisation = new Civilisation(Name.JAPANESE, GameVersion.AOK);
    public static readonly MONGOLS: Civilisation = new Civilisation(Name.MONGOLS, GameVersion.AOK);
    public static readonly PERSIANS: Civilisation = new Civilisation(Name.PERSIANS, GameVersion.AOK);
    public static readonly SARACENS: Civilisation = new Civilisation(Name.SARACENS, GameVersion.AOK);
    public static readonly TEUTONS: Civilisation = new Civilisation(Name.TEUTONS, GameVersion.AOK);
    public static readonly TURKS: Civilisation = new Civilisation(Name.TURKS, GameVersion.AOK);
    public static readonly VIKINGS: Civilisation = new Civilisation(Name.VIKINGS, GameVersion.AOK);

    public static readonly AZTECS: Civilisation = new Civilisation(Name.AZTECS, GameVersion.CONQUERORS);
    public static readonly HUNS: Civilisation = new Civilisation(Name.HUNS, GameVersion.CONQUERORS);
    public static readonly KOREANS: Civilisation = new Civilisation(Name.KOREANS, GameVersion.CONQUERORS);
    public static readonly MAYANS: Civilisation = new Civilisation(Name.MAYANS, GameVersion.CONQUERORS);
    public static readonly SPANISH: Civilisation = new Civilisation(Name.SPANISH, GameVersion.CONQUERORS);

    public static readonly ITALIANS: Civilisation = new Civilisation(Name.ITALIANS, GameVersion.FORGOTTEN);
    public static readonly INCAS: Civilisation = new Civilisation(Name.INCAS, GameVersion.FORGOTTEN);
    public static readonly INDIANS: Civilisation = new Civilisation(Name.INDIANS, GameVersion.FORGOTTEN);
    public static readonly MAGYARS: Civilisation = new Civilisation(Name.MAGYARS, GameVersion.FORGOTTEN);
    public static readonly SLAVS: Civilisation = new Civilisation(Name.SLAVS, GameVersion.FORGOTTEN);

    public static readonly BERBERS: Civilisation = new Civilisation(Name.BERBERS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly ETHIOPIANS: Civilisation = new Civilisation(Name.ETHIOPIANS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly MALIANS: Civilisation = new Civilisation(Name.MALIANS, GameVersion.AFRICAN_KINGDOMS);
    public static readonly PORTUGUESE: Civilisation = new Civilisation(Name.PORTUGUESE, GameVersion.AFRICAN_KINGDOMS);

    public static readonly BURMESE: Civilisation = new Civilisation(Name.BURMESE, GameVersion.RISE_OF_RAJAS);
    public static readonly KHMER: Civilisation = new Civilisation(Name.KHMER, GameVersion.RISE_OF_RAJAS);
    public static readonly MALAY: Civilisation = new Civilisation(Name.MALAY, GameVersion.RISE_OF_RAJAS);
    public static readonly VIETNAMESE: Civilisation = new Civilisation(Name.VIETNAMESE, GameVersion.RISE_OF_RAJAS);

    public static readonly BULGARIANS: Civilisation = new Civilisation(Name.BULGARIANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly CUMANS: Civilisation = new Civilisation(Name.CUMANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly LITHUANIANS: Civilisation = new Civilisation(Name.LITHUANIANS, GameVersion.DEFINITIVE_EDITION);
    public static readonly TATARS: Civilisation = new Civilisation(Name.TATARS, GameVersion.DEFINITIVE_EDITION);

    public static readonly BURGUNDIANS: Civilisation = new Civilisation(Name.BURGUNDIANS, GameVersion.LORDS_OF_THE_WEST);
    public static readonly SICILIANS: Civilisation = new Civilisation(Name.SICILIANS, GameVersion.LORDS_OF_THE_WEST);

    public static readonly BOHEMIANS: Civilisation = new Civilisation(Name.BOHEMIANS, GameVersion.DAWN_OF_THE_DUKES);
    public static readonly POLES: Civilisation = new Civilisation(Name.POLES, GameVersion.DAWN_OF_THE_DUKES);


    // DO NOT CHANGE THE ORDER OF ELEMENTS IN THIS ARRAY!!!
    // ONLY APPEND NEW CIVILISATIONS AT THE END!!!
    public static readonly ALL = [
        Civilisation.AZTECS,
        Civilisation.BERBERS,
        Civilisation.BRITONS,
        Civilisation.BURMESE,
        Civilisation.BYZANTINES,
        Civilisation.CELTS,
        Civilisation.CHINESE,
        Civilisation.ETHIOPIANS,
        Civilisation.FRANKS,
        Civilisation.GOTHS,
        Civilisation.HUNS,
        Civilisation.INCAS,
        Civilisation.INDIANS,
        Civilisation.ITALIANS,
        Civilisation.JAPANESE,
        Civilisation.KHMER,
        Civilisation.KOREANS,
        Civilisation.MAGYARS,
        Civilisation.MALAY,
        Civilisation.MALIANS,
        Civilisation.MAYANS,
        Civilisation.MONGOLS,
        Civilisation.PERSIANS,
        Civilisation.PORTUGUESE,
        Civilisation.SARACENS,
        Civilisation.SLAVS,
        Civilisation.SPANISH,
        Civilisation.TEUTONS,
        Civilisation.TURKS,
        Civilisation.VIETNAMESE,
        Civilisation.VIKINGS,
        Civilisation.BULGARIANS,
        Civilisation.CUMANS,
        Civilisation.LITHUANIANS,
        Civilisation.TATARS,
        Civilisation.BURGUNDIANS,
        Civilisation.SICILIANS,
        Civilisation.BOHEMIANS,
        Civilisation.POLES,
    ];

    public readonly gameVersion: GameVersion;

    private constructor(name: Name, gameVersion: GameVersion) {
        super(name);
        this.gameVersion = gameVersion;
    }

}

export default Civilisation;
export {Name};
