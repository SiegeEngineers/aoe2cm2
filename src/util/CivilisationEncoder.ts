import Civilisation from "../models/Civilisation";

export const CivilisationEncoder = {
    encodeCivilisationArray(civilisations: Civilisation[]): string {
        let encoded: number = 0;
        for (let i = 0; i < Civilisation.ALL.length; i++) {
            if (civilisations.includes(Civilisation.ALL[i])) {
                encoded += Math.pow(2, i);
            }
        }
        return CivilisationEncoder.toHexString(encoded);
    },

    decodeCivilisationArray(encoded: string): Civilisation[] {
        if (encoded === '7fffffffff') {
            const civilisations = [...Civilisation.ALL];
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        }
        try {
            const encodedNumber = parseInt(encoded, 16);
            const binaryString: string = encodedNumber.toString(2);
            const bits: string[] = binaryString.split('');
            const civilisations = [];
            for (let i = 0; i < bits.length; i++) {
                if (bits[i] === '1') {
                    civilisations.push(Civilisation.ALL[bits.length - 1 - i]);
                }
            }
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        } catch (e) {
            console.log("Could not decode value as hex string: ", encoded);
            return [];
        }
    },

    toHexString(input: number): string {
        return input.toString(16);
    }
};
