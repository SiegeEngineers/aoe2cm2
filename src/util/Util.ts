import Action from "../constants/Action";
import {DraftEvent} from "../types/DraftEvent";
import PlayerEvent from "../models/PlayerEvent";
import AdminEvent from "../models/AdminEvent";
import Turn from "../models/Turn";
import ActionType from "../constants/ActionType";
import {Validator} from "../models/Validator";
import {DraftsStore} from "../models/DraftsStore";
import Exclusivity from "../constants/Exclusivity";
import i18next from 'i18next';
import Player from "../constants/Player";
import DraftOption from "../models/DraftOption";
import {IDraftState} from "../types";
import LegacyPlayerEvent from "../models/LegacyPlayerEvent";
import Civilisation from "../models/Civilisation";

const CHARACTERS: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

export const Util = {
    notUndefined(...args: any[]): boolean {
        for (const arg of args) {
            if (arg === undefined) {
                return false;
            }
        }
        return true;
    },

    getIdFromUrl(): string {
        const match: RegExpMatchArray | null = window.location.pathname.match(/\/(?:draft|spectate)\/([A-Za-z]+)\/?.*/);
        if (match !== null) {
            return match[1];
        }
        alert('Could not get draft ID from url');
        return '';
    },

    isPick(action: Action): boolean {
        return (action === Action.PICK);
    },

    isNonglobalBan(turn: Turn): boolean {
        return (turn.action === Action.BAN && turn.exclusivity !== Exclusivity.GLOBAL);
    },

    isSnipe(action: Action): boolean {
        return (action === Action.SNIPE);
    },

    isPlayerEvent(event: DraftEvent): event is PlayerEvent {
        return (event as PlayerEvent).chosenOptionId !== undefined;
    },

    isLegacyPlayerEvent(event: unknown): event is PlayerEvent {
        return (event as LegacyPlayerEvent).civilisation !== undefined;
    },

    isAdminEvent(event: DraftEvent): event is AdminEvent {
        return !(event as AdminEvent).hasOwnProperty('chosenOptionId');
    },

    isCivilisation(draftOption: DraftOption): draftOption is Civilisation {
        return (draftOption as Civilisation).hasOwnProperty('gameVersion');
    },

    isCivilisationArray(draftOptions: DraftOption[]): draftOptions is Civilisation[] {
        for (const draftOption of draftOptions) {
            if (!Util.isCivilisation(draftOption)) {
                return false;
            }
        }
        return true;
    },

    isHidden(turn: Turn): boolean {
        return (turn.hidden);
    },

    getHiddenDraftOptionForActionType(actionType: ActionType): DraftOption {
        switch (actionType) {
            case ActionType.PICK:
                return DraftOption.HIDDEN_PICK;
            case ActionType.BAN:
                return DraftOption.HIDDEN_BAN;
            case ActionType.SNIPE:
                return DraftOption.HIDDEN_SNIPE;
            case ActionType.STEAL:
                return DraftOption.HIDDEN_STEAL;
            default:
                return DraftOption.HIDDEN;
        }
    },

    isTechnicalDraftOption(draftOption: DraftOption): boolean {
        switch (draftOption.id) {
            case DraftOption.HIDDEN.id:
            case DraftOption.HIDDEN_PICK.id:
            case DraftOption.HIDDEN_BAN.id:
            case DraftOption.HIDDEN_SNIPE.id:
            case DraftOption.HIDDEN_STEAL.id:
                return true;
        }
        return false;
    },

    isRandomDraftOption(id: string): boolean {
        return id.toUpperCase() === "RANDOM";
    },

    getRandomDraftOption(draftOptions: DraftOption[]): DraftOption {
        const maxIndex = draftOptions.length;
        const randomIndex = Math.floor(Math.random() * maxIndex);
        return draftOptions.splice(randomIndex, 1)[0];
    },

    setRandomDraftOptionIfNeeded(playerEvent: PlayerEvent, draftId: string,
                                 draftStore: DraftsStore, civilisationsList: DraftOption[], round: number = 100): PlayerEvent {
        if (round < 0) {
            return new PlayerEvent(playerEvent.player, playerEvent.actionType, DraftOption.HIDDEN.id, false, playerEvent.executingPlayer);
        }
        if (Util.isRandomDraftOption(playerEvent.chosenOptionId)) {
            const randomDraftOption = Util.getRandomDraftOption(civilisationsList);
            const playerEventForValidation = new PlayerEvent(playerEvent.player, playerEvent.actionType, randomDraftOption.id, true, playerEvent.executingPlayer);
            const errors = Validator.checkAllValidations(draftId, draftStore, playerEventForValidation);
            if (errors.length === 0) {
                playerEvent.chosenOptionId = randomDraftOption.id;
                playerEvent.isRandomlyChosen = true;
                return playerEvent;
            } else {
                return this.setRandomDraftOptionIfNeeded(playerEvent, draftId, draftStore, civilisationsList, round - 1);
            }
        }
        return playerEvent;
    },

    newDraftId(): string {
        let id: string = '';
        for (let i = 0; i < 5; i++) {
            id += CHARACTERS.charAt(Math.floor(Math.random() * CHARACTERS.length));
        }
        return id;
    },

    isValidPresetId(input: any): boolean {
        if (typeof input !== "string") {
            return false;
        }
        return input.match(new RegExp(`^[A-Za-z0-9_]+$`)) !== null;
    },

    randomChar(): string {
        return CHARACTERS.charAt(Math.floor(Math.random() * CHARACTERS.length));
    },

    buildValidationErrorMessage(data: any): string {
        if (data.hasOwnProperty('message')) {
            return data.message;
        }
        let message = i18next.t('validationFailed') + '\n';
        for (let validationError of data.validationErrors) {
            message += `\n${validationError}: ${i18next.t('errors.' + validationError)}`;
        }
        return message;
    },

    sanitizeDraftId(draftIdRaw: any) {
        if (typeof draftIdRaw === "string") {
            return draftIdRaw.replace(new RegExp(`[^${CHARACTERS}]`, 'g'), '_');
        }
        return '__invalid__';
    },

    sanitizeRole(roleRaw: Player) {
        if (roleRaw === Player.GUEST) {
            return Player.GUEST;
        }
        if (roleRaw === Player.HOST) {
            return Player.HOST;
        }
        return Player.NONE;
    },

    getAssignedRole(socket: SocketIO.Socket, roomHost: string, roomGuest: string): Player {
        let assignedRole: Player = Player.NONE;
        if (Object.keys(socket.rooms).includes(roomHost)) {
            assignedRole = Player.HOST;
        } else if (Object.keys(socket.rooms).includes(roomGuest)) {
            assignedRole = Player.GUEST;
        }
        return assignedRole;
    },

    isRequestFromLocalhost(req: any) {
        const ip = req.header('x-forwarded-for') || req.connection.remoteAddress;
        return ip === '::ffff:127.0.0.1' || ip === '::1';
    },

    getIconStyleFromLocalStorage(defaultIfError: string = 'units'): string {
        try {
            return localStorage.getItem('iconStyle') || defaultIfError;
        } catch (e) {
            return defaultIfError;
        }
    },

    writeIconStyleToLocalStorage(iconStyle: string) {
        try {
            localStorage.setItem('iconStyle', iconStyle);
        } catch (e) {
            // ignore
        }
    },

    transformDraftStateToCurrentFormat(input: IDraftState): IDraftState {
        for (let i = 0; i < input.events.length; i++) {
            const event = input.events[i];
            if (Util.isLegacyPlayerEvent(event)) {
                const legacyEvent = input.events[i] as unknown;
                input.events[i] = PlayerEvent.fromLegacy(legacyEvent as LegacyPlayerEvent);
            }
        }
        return input;
    },
};
