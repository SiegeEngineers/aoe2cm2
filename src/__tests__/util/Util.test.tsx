import {Util} from "../../util/Util";
import * as fs from "fs";
import path from "path";
import Civilisation from "../../models/Civilisation";
import DraftOption from "../../models/DraftOption";

it('sanitize handles common cases', () => {
    expect(Util.sanitizeDraftId('')).toEqual('');
    expect(Util.sanitizeDraftId('draftId')).toEqual('draftId');
    expect(Util.sanitizeDraftId('draft/Id')).toEqual('draft_Id');
    expect(Util.sanitizeDraftId('../../../../etc/shadow')).toEqual('____________etc_shadow');
    // @ts-ignore
    expect(Util.sanitizeDraftId(undefined)).toEqual('__invalid__');
});

it('getIdFromUrl works on draft url', () => {
    const draftId = 'abcdef';
    delete window.location;
    window.location = new URL('https://www.example.com/draft/' + draftId) as any;
    expect(Util.getIdFromUrl()).toEqual(draftId);
});

it('getIdFromUrl works on draft url with trailing slash', () => {
    const draftId = 'abcdef';
    delete window.location;
    window.location = new URL('https://www.example.com/draft/' + draftId + '/') as any;
    expect(Util.getIdFromUrl()).toEqual(draftId);
});

it('getIdFromUrl works on spectate url', () => {
    const draftId = 'abcdef';
    delete window.location;
    window.location = new URL('https://www.example.com/spectate/' + draftId) as any;
    expect(Util.getIdFromUrl()).toEqual(draftId);
});

it('getIdFromUrl works on spectate url with trailing slash', () => {
    const draftId = 'abcdef';
    delete window.location;
    window.location = new URL('https://www.example.com/spectate/' + draftId + '/') as any;
    expect(Util.getIdFromUrl()).toEqual(draftId);
});

describe('test isValidPresetId', () => {
    it.each([
        ["abcdef", true],
        ["ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_", true],
        ["abc123", true],
        ["Ã¤", false],
        ["", false],
        [undefined, false],
        ["https://example.com", false]
    ])("when the input is '%s'", (text: any, expected: boolean) => {
        expect(Util.isValidPresetId(text)).toBe(expected);
    });
});

it('Recognize Civilisation', () => {
    const civ = Civilisation.BRITONS;
    expect(Util.isCivilisation(civ)).toEqual(true);
});

it('Reject generic DraftOption as not a Civilisation', () => {
    const draftOption = new DraftOption('generic-draft-option');
    expect(Util.isCivilisation(draftOption)).toEqual(false);
});

it('Recognize Array of Civilisations', () => {
    const civs = [Civilisation.BRITONS, Civilisation.BERBERS, Civilisation.CUMANS];
    expect(Util.isCivilisationArray(civs)).toEqual(true);
});

it('Reject Array that contains non-Civilisation element', () => {
    const draftOptions = [Civilisation.BRITONS, new DraftOption('generic-draft-option'), Civilisation.CUMANS];
    expect(Util.isCivilisationArray(draftOptions)).toEqual(false);
});

it('RANDOM is not a technical draft option', () => {
    expect(Util.isTechnicalDraftOption(DraftOption.RANDOM)).toEqual(false);
});

it('Transforms legacy draft replay format', () => {
    const legacyDraft = JSON.parse(fs.readFileSync(path.join(__filename, '..', 'testfiles', 'legacy.json')).toString('utf8'));
    const transformed = Util.transformDraftStateToCurrentFormat(legacyDraft);
    expect(transformed).toMatchSnapshot();
});

it('Does not modify current draft replay format', () => {
    const draftInCurrentFormat = JSON.parse(fs.readFileSync(path.join(__filename, '..', 'testfiles', 'current.json')).toString('utf8'));
    const transformed = Util.transformDraftStateToCurrentFormat(draftInCurrentFormat);
    expect(transformed).toEqual(draftInCurrentFormat);
});
