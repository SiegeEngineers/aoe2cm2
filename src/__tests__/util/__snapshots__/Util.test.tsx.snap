// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transforms legacy draft replay format 1`] = `
Object {
  "events": Array [
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Aztecs",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 1395,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Berbers",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 2710,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Britons",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 3375,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Bulgarians",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 4512,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Burmese",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 5069,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Byzantines",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 6061,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Celts",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 6719,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Chinese",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 7622,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Cumans",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 8950,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Ethiopians",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 10311,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Franks",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": true,
      "offset": 11515,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Goths",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 12679,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Huns",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 13324,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "pick",
      "chosenOptionId": "Incas",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 18113,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "snipe",
      "chosenOptionId": "Byzantines",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 19752,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "snipe",
      "chosenOptionId": "Celts",
      "executingPlayer": "HOST",
      "isRandomlyChosen": false,
      "offset": 20646,
      "player": "HOST",
    },
    PlayerEvent {
      "actionType": "snipe",
      "chosenOptionId": "Huns",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": true,
      "offset": 22145,
      "player": "GUEST",
    },
    PlayerEvent {
      "actionType": "snipe",
      "chosenOptionId": "Goths",
      "executingPlayer": "GUEST",
      "isRandomlyChosen": false,
      "offset": 23261,
      "player": "GUEST",
    },
    Object {
      "action": "REVEAL_SNIPES",
      "offset": 25269,
      "player": "NONE",
    },
  ],
  "guestConnected": false,
  "guestReady": true,
  "hostConnected": false,
  "hostReady": true,
  "nameGuest": "Mohammad Shah",
  "nameHost": "Sultan Sanjar",
  "nextAction": 19,
  "preset": Object {
    "encodedCivilisations": "7ffffffff",
    "name": "LAT AMERICA INVITAT",
    "turns": Array [
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "5cf13f8d-fefd-41bf-8be7-b98b53377cec",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "c492ae70-b6e2-4d2f-bace-fa3e49ab8be3",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "e04f294f-9bd6-4907-afcf-bc877cff3cc5",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "a366d64b-18eb-4f88-ab33-b691f6ab2e77",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "d781a89a-894d-434c-be4c-59bde9c527f3",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "e66c6fe0-10bd-484f-8404-c99629167017",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "ae76b9fe-f89d-4f02-8605-fc06b2ad8d32",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "e92b9b5c-da2d-4e58-8391-82a9cf45f6b5",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "3332fa77-ddb9-4cb2-b3eb-35590eb97cb4",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "a54e3267-a043-46a7-8444-7d0d14629086",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "516cf7d9-be86-4c1e-8b2b-528747b6f8e8",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "609ae9ca-18c2-48bc-8009-c0409331c7ea",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "338d52ae-f078-42d4-98b4-169fea319586",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "PICK",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "c0beb8d8-55d4-46fd-b1d0-9e26f5238a54",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "SNIPE",
        "exclusivity": "GLOBAL",
        "hidden": true,
        "id": "621993a4-5f12-4a7b-933a-25c39a06f532",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "SNIPE",
        "exclusivity": "GLOBAL",
        "hidden": true,
        "id": "72899512-6df4-4cd2-a75d-1399a43bb72b",
        "parallel": false,
        "player": "HOST",
      },
      Object {
        "action": "SNIPE",
        "exclusivity": "GLOBAL",
        "hidden": true,
        "id": "6a50dac4-ec52-4878-830e-bb8766b26cca",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "SNIPE",
        "exclusivity": "GLOBAL",
        "hidden": true,
        "id": "c43342a1-b058-4936-80a8-479211e83101",
        "parallel": false,
        "player": "GUEST",
      },
      Object {
        "action": "REVEAL_SNIPES",
        "exclusivity": "GLOBAL",
        "hidden": false,
        "id": "70a00e51-4399-41c0-92ba-835262343e48",
        "parallel": false,
        "player": "NONE",
      },
    ],
  },
  "startTimestamp": 0,
}
`;
